File: Arrays/74-Search-a-2D-Matrix/Python/74-Search-a-2D-Matrix-Optimal.py
"""
Problem: 74. Search a 2D Matrix
Approach: Optimal - Binary Search on Virtual 1D Array

Treat the matrix as a sorted 1D array.
Convert the 1D index back to matrix coordinates:

row = mid // n
col = mid % n

Time Complexity: O(log(m * n))
Space Complexity: O(1)
"""

class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])

        low = 0
        high = m * n - 1

        while low <= high:
            mid = (low + high) // 2

            row = mid // n
            col = mid % n

            if matrix[row][col] == target:
                return True
            elif matrix[row][col] < target:
                low = mid + 1
            else:
                high = mid - 1

        return False

File: Arrays/74-Search-a-2D-Matrix/JavaScript/74-Search-a-2D-Matrix-Optimal.js
/**
 * Problem: 74. Search a 2D Matrix
 * Approach: Optimal - Binary Search on Virtual 1D Array
 *
 * Treat the matrix as a sorted 1D array.
 *
 * row = Math.floor(mid / n)
 * col = mid % n
 *
 * Time Complexity: O(log(m * n))
 * Space Complexity: O(1)
 */

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    const m = matrix.length;
    const n = matrix[0].length;

    let low = 0;
    let high = m * n - 1;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);

        const row = Math.floor(mid / n);
        const col = mid % n;

        if (matrix[row][col] === target) {
            return true;
        } else if (matrix[row][col] < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return false;
};